### ISSUE TITLE
Unable to install Nupic

### ISSUE THREAD

**Issue by Omairss** [numenta/nupic#1735]
I've been trying to install Nupic using "python setup.py install --user". I've installed all the dependencies and set the NUPIC and NTA variables. Previously i had managed to install it on OSX but i'm unable to get this to work. I've pasted the logs below:

```
-- Download successful.
-- Unpacking /home/omair/Workbox/nupic/extensions/core/build/release/nupic_core-4424626e3d6ca40e31fcaa35a93ae9592c3e6bed-linux64.tar.gz into /home/omair/Workbox/nupic/extensions/core/build/release/nupic_core-4424626e3d6ca40e31fcaa35a93ae9592c3e6bed-linux64...
-- Configuring done
-- Generating done
-- Build files have been written to: /home/omair/Workbox/nupic/build/scripts
Scanning dependencies of target py_support
[  7%] Building CXX object CMakeFiles/py_support.dir/extensions/py_support/NumpyVector.cpp.o
In file included from /usr/include/python2.7/numpy/ndarraytypes.h:1761:0,
                 from /usr/include/python2.7/numpy/ndarrayobject.h:17,
                 from /usr/include/python2.7/numpy/arrayobject.h:4,
                 from /home/omair/Workbox/nupic/extensions/py_support/NumpyVector.cpp:33:
/usr/include/python2.7/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
[ 14%] Building CXX object CMakeFiles/py_support.dir/extensions/py_support/PyArray.cpp.o
In file included from /usr/include/python2.7/numpy/ndarraytypes.h:1761:0,
                 from /usr/include/python2.7/numpy/ndarrayobject.h:17,
                 from /usr/include/python2.7/numpy/arrayobject.h:4,
                 from /home/omair/Workbox/nupic/extensions/py_support/PyArray.cpp:33:
/usr/include/python2.7/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
[ 21%] Building CXX object CMakeFiles/py_support.dir/extensions/py_support/PyHelpers.cpp.o
[ 28%] Building CXX object CMakeFiles/py_support.dir/extensions/py_support/PythonStream.cpp.o
Linking CXX static library libpy_support.a
[ 28%] Built target py_support
[ 35%] Swig source
/home/omair/Workbox/nupic/extensions/core/build/release/include/nupic/math/Math.hpp:521: Warning 325: Nested struct not currently supported (lexicographic ignored)
/home/omair/Workbox/nupic/extensions/core/build/release/include/nupic/math/Math.hpp:533: Warning 325: Nested struct not currently supported (less_value ignored)
/home/omair/Workbox/nupic/extensions/core/build/release/include/nupic/math/Math.hpp:545: Warning 325: Nested struct not currently supported (greater_value ignored)
/home/omair/Workbox/nupic/extensions/core/build/release/include/nupic/math/SparseBinaryMatrix.hpp:566: Warning 325: Nested struct not currently supported (lexicographic_order ignored)
/home/omair/Workbox/nupic/extensions/core/build/release/include/nupic/utils/LoggingException.hpp:36: Warning 401: Nothing known about base class 'Exception'. Ignored.
Scanning dependencies of target _algorithms
[ 42%] Building CXX object CMakeFiles/_algorithms.dir/temp/nupic/bindings/algorithmsPYTHON_wrap.cxx.o
In file included from /usr/include/python2.7/numpy/ndarraytypes.h:1761:0,
                 from /usr/include/python2.7/numpy/ndarrayobject.h:17,
                 from /usr/include/python2.7/numpy/arrayobject.h:4,
                 from /home/omair/Workbox/nupic/temp/nupic/bindings/algorithmsPYTHON_wrap.cxx:4387:
/usr/include/python2.7/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
In file included from /home/omair/Workbox/nupic/extensions/core/build/release/include/capnp/layout.h:37:0,
                 from /home/omair/Workbox/nupic/extensions/core/build/release/include/capnp/generated-header-support.h:31,
                 from /home/omair/Workbox/nupic/extensions/core/build/release/include/nupic/proto/RandomProto.capnp.h:7,
                 from /home/omair/Workbox/nupic/extensions/core/build/release/include/nupic/utils/Random.hpp:33,
                 from /home/omair/Workbox/nupic/extensions/core/build/release/include/nupic/math/ArrayAlgo.hpp:36,
                 from /home/omair/Workbox/nupic/extensions/core/build/release/include/nupic/math/SparseMatrix.hpp:38,
                 from /home/omair/Workbox/nupic/temp/nupic/bindings/algorithmsPYTHON_wrap.cxx:4368:
/home/omair/Workbox/nupic/extensions/core/build/release/include/kj/memory.h:213:28: error: expected ‘;’ at end of member declaration
/home/omair/Workbox/nupic/extensions/core/build/release/include/kj/memory.h:213:32: error: expected unqualified-id before ‘{’ token
/home/omair/Workbox/nupic/extensions/core/build/release/include/kj/memory.h:213:48: error: expected ‘;’ at end of member declaration
/home/omair/Workbox/nupic/extensions/core/build/release/include/kj/memory.h:214:36: error: expected ‘;’ at end of member declaration
/home/omair/Workbox/nupic/extensions/core/build/release/include/kj/memory.h:214:44: error: expected unqualified-id before ‘{’ token
/home/omair/Workbox/nupic/extensions/core/build/release/include/kj/memory.h:214:60: error: expected ‘;’ at end of member declaration
/home/omair/Workbox/nupic/extensions/core/build/release/include/kj/memory.h:215:29: error: expected ‘;’ at end of member declaration
/home/omair/Workbox/nupic/extensions/core/build/release/include/kj/memory.h:215:19: error: ‘kj::Own<T>&& kj::_::OwnOwn<T>::operator*()’ cannot be overloaded
/home/omair/Workbox/nupic/extensions/core/build/release/include/kj/memory.h:213:18: error: with ‘kj::Own<T>& kj::_::OwnOwn<T>::operator*()’
/home/omair/Workbox/nupic/extensions/core/build/release/include/kj/memory.h:215:34: error: expected unqualified-id before ‘{’ token
/home/omair/Workbox/nupic/extensions/core/build/release/include/kj/memory.h:215:58: error: expected ‘;’ at end of member declaration
/home/omair/Workbox/nupic/extensions/core/build/release/include/kj/memory.h:216:37: error: expected ‘;’ at end of member declaration
/home/omair/Workbox/nupic/extensions/core/build/release/include/kj/memory.h:216:25: error: ‘const kj::Own<T>&& kj::_::OwnOwn<T>::operator*() const’ cannot be overloaded
/home/omair/Workbox/nupic/extensions/core/build/release/include/kj/memory.h:214:24: error: with ‘const kj::Own<T>& kj::_::OwnOwn<T>::operator*() const’
/home/omair/Workbox/nupic/extensions/core/build/release/include/kj/memory.h:216:46: error: expected unqualified-id before ‘{’ token
/home/omair/Workbox/nupic/extensions/core/build/release/include/kj/memory.h:216:70: error: expected ‘;’ at end of member declaration
In file included from /home/omair/Workbox/nupic/extensions/core/build/release/include/nupic/proto/RandomProto.capnp.h:7:0,
                 from /home/omair/Workbox/nupic/extensions/core/build/release/include/nupic/utils/Random.hpp:33,
                 from /home/omair/Workbox/nupic/extensions/core/build/release/include/nupic/math/ArrayAlgo.hpp:36,
                 from /home/omair/Workbox/nupic/extensions/core/build/release/include/nupic/math/SparseMatrix.hpp:38,
                 from /home/omair/Workbox/nupic/temp/nupic/bindings/algorithmsPYTHON_wrap.cxx:4368:
/home/omair/Workbox/nupic/extensions/core/build/release/include/capnp/generated-header-support.h: In constructor ‘constexpr capnp::_::RawBrandedSchema::Binding::Binding(uint8_t, uint16_t, const capnp::_::RawBrandedSchema*)’:
/home/omair/Workbox/nupic/extensions/core/build/release/include/capnp/generated-header-support.h:82:27: error: uninitialized member ‘capnp::_::RawBrandedSchema::Binding::<anonymous>’ in ‘constexpr’ constructor
/home/omair/Workbox/nupic/extensions/core/build/release/include/capnp/generated-header-support.h: In constructor ‘constexpr capnp::_::RawBrandedSchema::Binding::Binding(uint8_t, uint16_t, uint64_t, uint16_t)’:
/home/omair/Workbox/nupic/extensions/core/build/release/include/capnp/generated-header-support.h:86:29: error: uninitialized member ‘capnp::_::RawBrandedSchema::Binding::<anonymous>’ in ‘constexpr’ constructor
/home/omair/Workbox/nupic/extensions/core/build/release/include/capnp/generated-header-support.h: In constructor ‘constexpr capnp::_::RawBrandedSchema::Binding::Binding(uint8_t, uint16_t, uint16_t)’:
/home/omair/Workbox/nupic/extensions/core/build/release/include/capnp/generated-header-support.h:89:55: error: uninitialized member ‘capnp::_::RawBrandedSchema::Binding::<anonymous>’ in ‘constexpr’ constructor
CMakeFiles/_algorithms.dir/build.make:62: recipe for target 'CMakeFiles/_algorithms.dir/temp/nupic/bindings/algorithmsPYTHON_wrap.cxx.o' failed
make[2]: *** [CMakeFiles/_algorithms.dir/temp/nupic/bindings/algorithmsPYTHON_wrap.cxx.o] Error 1
CMakeFiles/Makefile2:60: recipe for target 'CMakeFiles/_algorithms.dir/all' failed
make[1]: *** [CMakeFiles/_algorithms.dir/all] Error 2
Makefile:113: recipe for target 'all' failed
make: *** [all] Error 2
Unable to build the library!
```

**Comment by rhyolight** [comment]
@Omairss What are your cmake and clang version?

    clang --version
    cmake --version

**Comment by Omairss** [comment]
omair@omair-workstation:~/Workbox/nupic$ make --version
GNU Make 4.0
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

omair@omair-workstation:~/Workbox/nupic$ clang --version
Ubuntu clang version 3.5.0-4ubuntu2 (tags/RELEASE_350/final) (based on LLVM 3.5.0)
Target: x86_64-pc-linux-gnu
Thread model: posix

omair@omair-workstation:~/Workbox/nupic$ clang --version
Ubuntu clang version 3.5.0-4ubuntu2 (tags/RELEASE_350/final) (based on LLVM 3.5.0)
Target: x86_64-pc-linux-gnu
Thread model: posix

I'm running Ubuntu Desktop 14.10

**Comment by rhyolight** [comment]
@Omairss You have me clang twice! :stuck_out_tongue_winking_eye: Can you give me your cmake version?

**Comment by breznak** [comment]
> Also exactly what OS X version are you using? 

he's using the best OSX :trollface: :wink: 

> I'm running Ubuntu Desktop 14.10

**Comment by rhyolight** [comment]
Oh I see. I got confused by the first report mentioning OS X. Make more sense now. 

@Omairss I also need to know that your NuPIC checkout is up-to-date. Can you make sure you have the latest code? (If not, try installing again)

**Comment by Omairss** [comment]
Yep, atleast at that point of time the checkout was very much up to date. Either way, i tired everything from <code> git pull </code> to trying out different versions off gcc and g++ compilers. finally, i ended up reinstalling the OS all over again, and now there seems to be a problem with AWS?

Here are the logs:

<code>
omair@omair-workstation:~/Workbox/nupic$ python setup.py install --user-- The CXX compiler identification is GNU 4.9.1
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
Python 2.7.8
doing local build
-- Attempting to fetch nupic.core binaries from https://s3-us-west-2.amazonaws.com/artifacts.numenta.org/numenta/nupic.core/nupic_core-4424626e3d6ca40e31fcaa35a93ae9592c3e6bed-linux32.tar.gz and save to /home/omair/Workbox/nupic/extensions/core/build/release/nupic_core-4424626e3d6ca40e31fcaa35a93ae9592c3e6bed-linux32.tar.gz...
CMake Error at CMakeLists.txt:232 (message):
  Failed to download nupic.core tarball from
  https://s3-us-west-2.amazonaws.com/artifacts.numenta.org/numenta/nupic.core/nupic_core-4424626e3d6ca40e31fcaa35a93ae9592c3e6bed-linux32.tar.gz!
  Ensure you have an internet connection and that the remote tarball exists.


-- Configuring incomplete, errors occurred!
See also "/home/omair/Workbox/nupic/build/scripts/CMakeFiles/CMakeOutput.log".
Unable to generate build scripts!
</code>

Thanks!

**Comment by Omairss** [comment]
I cannot seem to get to the bottom of this. :( Here's the output of python setup.py. I really don't know what i could be doing wrong. I'm running cmake version 2.8.12.2 and GNU Make 4.0 and Ubuntu clang version 3.5.0-4ubuntu2.



omair@omair-workstation:~/Workbox/nupic$ sudo python setup.py install
-- The CXX compiler identification is GNU 4.7.4
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
Python 2.7.8
doing local build
-- Target nupic.core package already exists at /home/omair/Workbox/nupic/extensions/core/build/release/nupic_core-4424626e3d6ca40e31fcaa35a93ae9592c3e6bed-linux64.tar.gz.
-- Unpacking /home/omair/Workbox/nupic/extensions/core/build/release/nupic_core-4424626e3d6ca40e31fcaa35a93ae9592c3e6bed-linux64.tar.gz into /home/omair/Workbox/nupic/extensions/core/build/release/nupic_core-4424626e3d6ca40e31fcaa35a93ae9592c3e6bed-linux64...
-- Configuring done
-- Generating done
-- Build files have been written to: /home/omair/Workbox/nupic/build/scripts
Scanning dependencies of target py_support
[  7%] Building CXX object CMakeFiles/py_support.dir/extensions/py_support/NumpyVector.cpp.o
In file included from /home/omair/.local/lib/python2.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1804:0,
                 from /home/omair/.local/lib/python2.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:17,
                 from /home/omair/.local/lib/python2.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from /home/omair/Workbox/nupic/extensions/py_support/NumpyVector.cpp:33:
/home/omair/.local/lib/python2.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
[ 14%] Building CXX object CMakeFiles/py_support.dir/extensions/py_support/PyArray.cpp.o
In file included from /home/omair/.local/lib/python2.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1804:0,
                 from /home/omair/.local/lib/python2.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:17,
                 from /home/omair/.local/lib/python2.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from /home/omair/Workbox/nupic/extensions/py_support/PyArray.cpp:33:
/home/omair/.local/lib/python2.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
[ 21%] Building CXX object CMakeFiles/py_support.dir/extensions/py_support/PyHelpers.cpp.o
[ 28%] Building CXX object CMakeFiles/py_support.dir/extensions/py_support/PythonStream.cpp.o
Linking CXX static library libpy_support.a
[ 28%] Built target py_support
[ 35%] Swig source
make[2]: execvp: ../../external/linux64/bin/swig: Permission denied
CMakeFiles/_algorithms.dir/build.make:53: recipe for target '../../temp/nupic/bindings/algorithmsPYTHON_wrap.cxx' failed
make[2]: *** [../../temp/nupic/bindings/algorithmsPYTHON_wrap.cxx] Error 127
CMakeFiles/Makefile2:60: recipe for target 'CMakeFiles/_algorithms.dir/all' failed
make[1]: *** [CMakeFiles/_algorithms.dir/all] Error 2
Makefile:113: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by rhyolight** [comment]
Does "python setup.py install --user" work?

---------
Matt Taylor
OS Community Flag-Bearer
Numenta

**Comment by Omairss** [comment]
Nope! Tried that too. I get the same error. :/

**Comment by Omairss** [comment]
I finally managed to install it! I changed the default cc and c++ compiler to clang and clang++ using update-alternatives. To anyone facing the same issue, I used the instructions over here http://askubuntu.com/questions/26498/choose-gcc-and-g-version

Man, I don't think I've spent so much time on my girlfriend even, though I'm sure it's worth it. Time for some CLA goodness! Thanks for your time, Matt and breznak!
